ARCHITECTURE.md

```diff
@@ -0,0 +1,170 @@
+
+## kURL Installer
+
+### Entrypoints
+
+| Script | Description |
+| ------ | ----------- |
+| install.sh | Initialize and upgrade a kURL cluster. |
+| join.sh    | Join additional nodes to the cluster. |
+| upgrade.sh | Utility script for upgrading additional nodes. |
+| tasks.sh   | For various tasks including re-generating the join script and loading images from an air-gap bundle. |
+
+### kURL Directory Structure
+
+Exists on the server at `/var/lib/kurl`.
+
+| Directory | Description |
+| ----------| ----------- |
+| addons/    | Stores addons, including images and assets. |
+| bin/       | Stores utility binaries. |
+| krew/      | Has some plugins we use including support-bundle and preflight. |
+| kurlkinds/ | Contains cluster.kurl.sh CRD |
+| kustomize/ | Scripts output  |
+| packages/  | Host packages |
+| shared/    | Everything else |
+
+### Flow Chart
+
+#### Installation Lifecycle
+
+![kURL sh Installation Lifecycle](https://user-images.githubusercontent.com/371319/166290254-f73a4232-5227-45b5-a673-cc742e626e57.png)
+
+#### Join Lifecycle
+
+![kURL sh Join Lifecycle](https://user-images.githubusercontent.com/371319/166290269-b6451622-2890-491e-98cc-6f20997c8b13.png)
+
+## Add-ons
+
+Add-ons are components that make up a kURL cluster.
+
+### Categories
+
+1. Kubernetes distribution - Kubeadm, RKE2, K3s
+1. CRI - Docker or Containerd
+1. CNI - Weave or Antrea
+1. CSI - Longhorn, Rook, OpenEBS
+1. Ingress - Contour
+1. Misc. - Prometheus, KOTS, Velero...
+
+### Directory Structure
+
+| Directory | Description |
+| --------- | ----------- |
+| addons/[addon]/[version]/Manifest            | Manifest of assets, host packages and container images |
+| addons/[addon]/[version]/install.sh          | Entrypoint to the add-on installation script |
+| addons/[addon]/[version]/host-preflight.yaml | Troubleshoot.sh preflight spec |
+| addons/[addon]/[version]/assets/             | Runtime assets |
+| addons/[addon]/[version]/images/             | Runtime images |
+| addons/[addon]/[version]/[distro-version]/   | Runtime host packages for each supported Linux OS  |
+
+### Lifecycle
+
+#### addon_fetch
+
+Fetch the add-on package from the object store or from the air-gap bundle.
+This step is typically skipped if the add-on version has not changed since the previous run.
+
+#### addon_load
+
+Load (bash source) the install.sh script.
+
+#### addon_preflights
+
+Run the Troubleshoot.sh preflight spec from host-preflight.yaml.
+
+#### addon_pre_init
+
+Operations that are performed in the pre-init script include installing host packages or Kustomizing the Kubernetes distribution.
+
+#### addon_install
+
+Kubectl apply this add-on to the cluster.
+This step is typically skipped if the add-on version and configuration has not changed since the previous run.
```

Comment:
we should mention what we do upon skipping (we attempt to run `addon_already_applied`)

Reply:
TIL